### PR TITLE
planner: fix optimizer trace may miss operators in some cases

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/31811

Problem Summary:

### What is changed and how it works?

fix optimizer trace may miss operators in some cases

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/physical_plans.go [hunk 1]
```diff
@@ -223,6 +224,30 @@ func (p *PhysicalTableReader) ExtractCorrelatedCols() (corCols []*expression.Cor
 	return corCols
 }
 
+func (p *PhysicalTableReader) buildPlanTrace() *tracing.PlanTrace {
```
Comment (time-and-fate) [inline planner/core/physical_plans.go:227]: Seems we should also consider `PhysicalIndexReader`, `PhysicalIndexLookUpReader`, `PhysicalIndexMergeReader`...

File: planner/core/physical_plans.go [hunk 2]
```diff
@@ -223,6 +224,30 @@ func (p *PhysicalTableReader) ExtractCorrelatedCols() (corCols []*expression.Cor
...
+	rp := p.basePhysicalPlan.buildPlanTrace()
+	if p.tablePlan != nil {
+		rp.Children = append(rp.Children, p.tablePlan.buildPlanTrace())
+	}
+	return rp
+}
+
+func (p *PhysicalTableReader) appendChildCandidate(op *physicalOptimizeOp) {
```
Comment (time-and-fate) [inline planner/core/physical_plans.go:235]: ditto.

**Review by time-and-fate — COMMENTED** [review]
Can we simplify the logic a bit? I find it's not easy to understand them.

Some simple questions I have at first sight:
- What's the expected result in the `PhysicalOptimizeTracer`?
- What's the relationship between `PhysicalPlan.appendChildCandidate`, `physicalOptimizeOp.appendCandidate`, `appendCandidate` and `PhysicalOptimizeTracer.AppendCandidate`?
- When should a `CandidatePlanTrace` have a `MappingLogicalPlan`?
- When should a `PlanTrace` have `Children` or `ChildrenID`?

File: planner/core/physical_plans.go (hunk 1 above)
Comment (Yisaer) [reply planner/core/physical_plans.go:228]: we will finish other operator in future pr in order to keep each pr simple.

**Review by time-and-fate — APPROVED** [review]

File: util/tracing/opt_trace.go [hunk 3]
```diff
@@ -178,48 +184,18 @@ type CandidatePlanTrace struct {
...
-	}
-	for logicalKey, pps := range tracer.State {
-		for _, pp := range pps {
-			c := newCandidatePlanTrace(pp, logicalKey, bestKeys)
-			if c.Selected {
-				sCandidates = append(sCandidates, c)
-			} else {
-				dCandidates = append(dCandidates, c)
+	for _, candidate := range tracer.Candidates {
+		for _, plan := range tracer.Final {
```
Comment (hawkingrei) [inline util/tracing/opt_trace.go:193]: loop-in-loop can be replaced with two loop and check an hashset to find the exist.

File: planner/core/physical_plan_trace_test.go [hunk 4]
```diff
@@ -105,14 +90,93 @@ func checkList(d []string, s []string) bool {
...
+		16: {},
+		10: {},
+		9:  {},
+		8:  {},
+	}
+	asserted := true
+	for _, c := range otrace.Candidates {
+		tp, ok := elements[c.ID]
+		if !ok || tp != c.TP {
+			asserted = false
```
Comment (hawkingrei) [inline planner/core/physical_plan_trace_test.go:167]: t.fail ?

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**